package antixtransfer;

// node -> master initial connection message
message connect_init_node {
	required string ip_addr = 1;
	required string announce_port = 2;
}

// master -> node/client response upon initial connect
message connect_init_response {
	required int32 id = 1;
	required double world_size = 2;
	required int32 sleep_time = 3;
	// only relevant for node
	optional int32 puck_amount = 4;

	// more... same as MasterServerClient/NodeInitialization?
}

message node_node_sync {
	required int32 id = 1;
}

message node_master_sync {
	required int32 my_id = 1;
	required int32 heard_id = 2;
}

// used to have synchronous turns
message node_master_done {
	required int32 my_id = 1;
}

// master -> node/client list of nodes
message Node_list {
	message Node {
		required string ip_addr = 1;
		required string left_port = 2;
		required string announce_port = 3;
		required int32 id = 4;
		// optional as we only set this right before sending out node list
		optional double x_offset = 6;
	}
	repeated Node node = 1;
}

message connect {
	enum connectionType {
	     CLIENT = 0;
	     SERVER = 1;
	}

	required connectionType type = 1 [default = CLIENT];
}

message MasterServerSlaveServerIntialization {
	required int32 robotsAmt  = 1;
	required int32 teamAmt = 2;
	required int32 minRobotID = 3;
	required int32 maxRobotID = 4;
	required int32 puckAmt = 5;
	required double visionrange = 6; 
	required double fieldofview = 7;	

	message Home {
		required int32 homeRadius = 6;
		required int32 homeX = 7;
		required int32 homeY = 8;
		required int32 homeID = 9;
	}

	repeated Home home = 8;
	optional string leftServersIP = 9;
	optional string rightServersIP = 10;
}

message MasterServerClientInitialization {
	required double visionrange = 1;
	required double fieldofview = 2;	
	required double serverwidth = 3;
	required double serverheight = 4;
	required double homeradius = 5;
	// id assigned to the client
	required int32 id = 6;
	required int32 sleep_time = 7;

	message HomeLocations {
		required int32 team = 1;
		required double x = 2;
		required double y = 3;
		required string server = 4;
	}

	repeated HomeLocations homelocations = 8;
}	

message ClientSlaveServerStartup{
	required int32 numberofrobots = 1;
}

message SlaveServerClientInitialization{
	message Robot {
		required int32 robotID = 1;
		required double robotX = 2;
		required double robotY = 3;
		required double robotfacing = 4;
		required int32 robotTeam = 5;
	}

	repeated Robot robot = 1;
}

message RobotRequestMap {
	required int32 robotID = 1;
	required double robotX = 2;
	required double robotY = 3;
	required double robotfacing = 4;
}

message SendMap {
	message Robot {
		required int32 team = 1;
		required int32 id = 2;
		required double x = 3;
		required double y = 4;
	}

	message Puck {
		required double x = 1;
		required double y = 2;
		required bool held = 3;
	}

	repeated Robot robot = 1;
	repeated Puck puck = 2;
}

// potentially used for redesign
message SendMap2 {
	message Robot {
		required double x = 1;
		required double y = 2;
		optional int32 puck_id = 3;
		optional bool puck_action = 4;
	}
	repeated Robot robot = 1;
}

message ServerRequestMap {
	message MapLocation {
		required double mapX = 1;
		required double mapY = 2;
	}

	repeated MapLocation maplocation = 1;
}

message RequestMovement {
	required int32 robotID = 1;
	required double robotx = 2;
	required double roboty = 3;
}

message MovementUpdate {
	required bool robotmoved  = 1;
}

message RequestPuckHandling {
	required double puckX = 1;
	required double puckY = 2;

	enum ActionType {
	     PICKUP = 1;
	     DROP = 2;	
	}

	required ActionType action = 3;
}

message PuckActionUpdate {
	required bool actionSuccess  = 1;
}

message RequestRobotTransfer {
	required int32 id = 1;
	required int32 team = 2;
	required double x = 3;
	required double y = 4;
	required double a = 5;
	required double v = 6;
	required double w = 7;
	required bool has_puck = 8;
}

// not needed if send message over PUB socket?
message TransferUpdate {
	required bool success = 1;
}

message control_message {
	enum Type {
		ADD_BOT = 1;
		SENSE = 2;
		SETSPEED = 3;
		PICKUP = 4;
		DROP = 5;
	}

	message Robot {
		required int32 id = 1;
		optional double v = 2;
		optional double w = 3;
	}

	// all robots in message with type TYPE get the action done to them

	required int32 team = 1;
	required Type type = 2;
	repeated Robot robot = 3;
}
